### PR TITLE
update the project archive sync support for py2 compatability

### PR DESCRIPTION
**Pull request by ryanpetrello**
playbook runs in AWX are still supported in python2, so this adds
support for archive-based project syncs in py2

cc @philipsd6

### CODE REVIEW

File: awx/playbooks/action_plugins/project_archive.py [hunk 1]
```diff
@@ -26,14 +32,15 @@ def run(self, tmp=None, task_vars=None):
             archive = zipfile.ZipFile(src)
             get_filenames = archive.namelist
             get_members = archive.infolist
-        except zipfile.BadZipFile:
-            archive = tarfile.open(src)
+        except BadZipFile:
+            try:
+                archive = tarfile.open(src)
+            except tarfile.ReadError:
```
Comment (ryanpetrello) [inline awx/playbooks/action_plugins/project_archive.py:38]: This lack of indentation was actually a legitimate bug we just missed in code review.

File: awx/playbooks/action_plugins/project_archive.py [hunk 2]
```diff
@@ -2,15 +2,21 @@
...
 
 from ansible.plugins.action import ActionBase
 from ansible.utils.display import Display
 
 display = Display()
 
+try:
+    from zipfile import BadZipFile
+except ImportError:
+    from zipfile import BadZipfile as BadZipFile  # py2 compat
```
Comment (ryanpetrello) [inline awx/playbooks/action_plugins/project_archive.py:18]: Booooooooooooo python2

File: awx/playbooks/action_plugins/project_archive.py [hunk 3]
```diff
@@ -62,10 +69,19 @@ def run(self, tmp=None, task_vars=None):
             try:
                 is_dir = member.is_dir()
             except AttributeError:
-                is_dir = member.isdir()
+                try:
+                    is_dir = member.isdir()
+                except AttributeError:
+                    is_dir = member.filename[-1] == '/'  # py2 compat for ZipInfo
```
Comment (ryanpetrello) [inline awx/playbooks/action_plugins/project_archive.py:75]: python2's `zipfile` doesn't have this, but this is what py3 does, so I just copied it here <shrug>

File: awx/playbooks/action_plugins/project_archive.py [hunk 4]
```diff
@@ -62,10 +69,19 @@ def run(self, tmp=None, task_vars=None):
...
 
             if is_dir:
-                os.makedirs(dest, exist_ok=True)
+                try:
+                    os.makedirs(dest)
```
Comment (ryanpetrello) [inline awx/playbooks/action_plugins/project_archive.py:79]: poor man's `exist_ok=True` (which also doesn't exist in py2)

**Review by kdelee — APPROVED** [review]
:+1: fixes the tests that were failing, works for me!

**Comment by philipsd6** [comment]
@ryanpetrello I had the understanding that AWX was solidly Python3 at this point, and even if a project was associated with a Py2 venv, the project sync would be done with the default Py3, no? Seems unreasonable for the built ins to not be able to depend on the main Python version! In any case, as my understanding was apparently wrong, thanks for catching this and fixing it!

P.S. I'm **so** tired of writing backwards compatible Python... Sometimes I even get bit when using `{k: v for k, v in ...}` instead of `dict((k, v) for k, v in ...)`

**Comment by ryanpetrello** [comment]
Hey @philipsd6,

Hey no big deal.  I'm a maintainer and I forgot too 😂.

When a project update runs, it uses the default "Ansible" virtualenv, and we still technically support systems (like RHEL7) that have python2 as their default system Python for _playbook execution_.

The Python code that AWX itself runs in _its_ various processes is all python3 only, though.

**Pull request merged**